ext/test/zpages/tracez_processor_test.cc

```diff
@@ -0,0 +1,586 @@
+
+  EXPECT_TRUE(ContainsNames(span_names, running, 0, 1)); // s0
+  EXPECT_TRUE(ContainsNames(span_names, running, 4)); // + s
+  EXPECT_TRUE(ContainsNames(span_names, completed, 1, 3, true)); // s2 s1
+  EXPECT_EQ(running.size(), 2);
+  EXPECT_EQ(completed.size(), 2);
+
+  // End remaining Spans
+  span_vars[0]->End();
+  span_vars[4]->End();
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_TRUE(ContainsNames(span_names, completed, 0, 1)); // s0
+  EXPECT_TRUE(ContainsNames(span_names, completed, 4)); // s
+  EXPECT_EQ(running.size(), 0);
+  EXPECT_EQ(completed.size(), 2);
+
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_EQ(running.size(), 0);
+  EXPECT_EQ(completed.size(), 0);
+}
+
+
+/*
+ * Test if multiple spans move from running to completed at expected times,
+ * running/completed spans are split. Ensure correct behavior even when
+ * multiple spans are discarded, even when span starting and ending is
+ * non-sequential. Only new completed spans are stored.
+*/
+TEST_F(TracezProcessor, MultipleSpansOuterSplitNewOnly) {
+  for (const auto &name : span_names) span_vars.push_back(tracer->StartSpan(name));
+
+  // End last span
+  span_vars[4]->End();
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_TRUE(ContainsNames(span_names, running, 0, 4, true)); // s0 s2 s1 s1
+  EXPECT_TRUE(ContainsNames(span_names, completed, 4, 5, true)); // s
+  EXPECT_EQ(running.size(), 4);
+  EXPECT_EQ(completed.size(), 1);
+
+  // End first span
+  span_vars[0]->End();
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_TRUE(ContainsNames(span_names, running, 1, 4, true)); // s2 s1 s1
+  EXPECT_TRUE(ContainsNames(span_names, completed, 0, 1, true)); // s0
+  EXPECT_EQ(running.size(), 3);
+  EXPECT_EQ(completed.size(), 1);
+
+  // End remaining middle spans
+  for (int i = 1; i < 4; i++) span_vars[i]->End();
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_TRUE(ContainsNames(span_names, completed, 1, 4, true)); // s2 s1 s1
+  EXPECT_EQ(running.size(), 0);
+  EXPECT_EQ(completed.size(), 3);
+
+  UpdateSpans(processor, completed, running, true);
+
+  EXPECT_EQ(running.size(), 0);
+  EXPECT_EQ(completed.size(), 0);
+}
+
+
+/*
+ * Test for ForceFlush and Shutdown code coverage, which do nothing.
+ */
+TEST_F(TracezProcessor, FlushShutdown) {
+  auto pre_running_sz = running.size();
+  auto pre_completed_sz = completed.size();
+
+  processor->ForceFlush();
+  processor->Shutdown();
+
+  UpdateSpans(processor, completed, running);
+  
+  EXPECT_EQ(pre_running_sz, running.size());
+  EXPECT_EQ(pre_completed_sz, completed.size());
+}
+
+
+/*
+ * Test for thread safety when many spans start at the same time.
+ */
+TEST_F(TracezProcessor, RunningThreadSafety) {
+  std::vector<opentelemetry::nostd::unique_ptr<opentelemetry::trace::Span>> spans1;
+  std::vector<opentelemetry::nostd::unique_ptr<opentelemetry::trace::Span>> spans2;
+
+  std::thread start1(StartManySpans, std::ref(spans1), tracer, 5);
```

Comment:
You can use much higher numbers here. 5 spans might be started well before the second thread gets a chance to start. Consider `5000` instead.

Reply:
Would `2000` spans work? I tried `5000` and the tests for the TraceZ processor alone were ~24sec. `2000` has a test time of 2-3sec, while still being much larger than 5.

Reply:
Oh sorry, I didn't expect it to take that long. I think even 500 spans would be fine then to assure we actually run concurrently for a reasonable time.